{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Machine learning prediction of trophic modes\n",
    "In this notebook, we'll predict whether phaeocystisn is phototrophic, heterotrophic or mixotrophic in a given sample.\n",
    "\n",
    "### Setup\n",
    "First, we'll import the necessary scripts and load the training data from [github](https://github.com/armbrustlab/trophic-mode-ml) and [zenodo](https://zenodo.org/record/4425690#.ZHWzKC9ByLf).\n",
    "These are stored in the `data` folder.\n",
    "Then, we'll run the model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import requests\n",
    "import numpy as np\n",
    "import re\n",
    "import sys\n",
    "import warnings\n",
    "from pandas.errors import PerformanceWarning\n",
    "warnings.filterwarnings('ignore', category=PerformanceWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18720\n",
      "7612\n"
     ]
    }
   ],
   "source": [
    "# Read transcripts annotated to Phaeocystis (from the output of the mtx_taxonomy.ipynb script)\n",
    "phaeocystis_transcripts = pd.read_csv(\"../data/annotation/taxonomy_eukprot/130/genus_bins/Phaeocystis_transcriptome_bin.csv\")\n",
    "print(len(phaeocystis_transcripts))\n",
    "# Filter out transcripts annotated to Phaeocystis globosa with a percent identity above 80\n",
    "phaeocystis_transcripts = phaeocystis_transcripts[(phaeocystis_transcripts[\"p_ident\"] > 0.8) & (phaeocystis_transcripts[\"Name_to_Use\"] == \"Phaeocystis globosa\")]\n",
    "print(len(phaeocystis_transcripts))\n",
    "\n",
    "# Keep only the transcript IDs\n",
    "phaeocystis_transcripts = phaeocystis_transcripts[\"query_id\"].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "165198\n",
      "5771\n"
     ]
    }
   ],
   "source": [
    "# Read in the functional annotation data\n",
    "functional_annotation = pd.read_table('../data/annotation/functional/130/functional_annotation.emapper.annotations', )\n",
    "# Cut off weird characters from the transcript names\n",
    "functional_annotation['#query'] = functional_annotation['#query'].str.split(\".\", n=1, expand=True)[0]\n",
    "# Rename the query_id column\n",
    "functional_annotation.rename(columns={'#query': 'transcript_id'}, inplace=True)\n",
    "print(len(functional_annotation))\n",
    "# Only keep the rows that are in the Phaeocystis transcriptome\n",
    "functional_annotation = functional_annotation[functional_annotation['transcript_id'].isin(phaeocystis_transcripts)]\n",
    "print(len(functional_annotation))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the quantification data\n",
    "tpm = pd.read_csv('../data/quantification/130_tpm.csv')\n",
    "\n",
    "# Rename the target_id column\n",
    "tpm.rename(columns={'target_id': 'transcript_id'}, inplace=True)\n",
    "\n",
    "# Change into long format\n",
    "tpm = tpm.melt(id_vars=['transcript_id'], var_name='sample', value_name='TPM')\n",
    "\n",
    "# Only keep the rows that are in the Phaeocystis transcriptome\n",
    "tpm = tpm[tpm['transcript_id'].isin(phaeocystis_transcripts)]\n",
    "\n",
    "## Generate Phaeocystis tpm sums per sample\n",
    "phaeo_tpm_sums = tpm.groupby('sample')['TPM'].sum().reset_index()\n",
    "\n",
    "## Rename columns\n",
    "phaeo_tpm_sums.columns = ['sample', 'Phaeocystis_total_tpm']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['sample', 'StationPrefix', 'StationSuffix', 'Latitude', 'Longitude',\n",
      "       'Date', 'day_moment', 'day_length', 'Temperature', 'Salinity',\n",
      "       'Conductivity', 'Depth', 'Oxygen', 'Fluorescence', 'NH4', 'NO2', 'NO3',\n",
      "       'NOX', 'PO4', 'Si', 'TEP', 'sea_surface_height_above_sea_level',\n",
      "       'surface_baroclinic_sea_water_velocity', 'PP'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# Read in the environmental data\n",
    "env_data = pd.read_csv('../data/samples_env.csv', index_col=0)\n",
    "# Read in primary production data\n",
    "PP_data = pd.read_csv('../data/raw/LabSTAF/labstaf_combined_data.csv')\n",
    "## Combine Station and Sample columns into one station_sample column\n",
    "PP_data['Station'] = PP_data['Station'].astype(str) + '_' + PP_data['Sample'].astype(str)\n",
    "\n",
    "\n",
    "# Merge PP_data with env and transcript data\n",
    "env_data = env_data.merge(PP_data[['Station', 'PP']], on=['Station'], how='inner')\n",
    "\n",
    "# Rename station column\n",
    "env_data.rename(columns={'Station': 'sample'}, inplace=True)\n",
    "\n",
    "print(env_data.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>PFAMs</th>\n",
       "      <th>sample</th>\n",
       "      <th>TPM</th>\n",
       "      <th>StationPrefix</th>\n",
       "      <th>StationSuffix</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>Longitude</th>\n",
       "      <th>Date</th>\n",
       "      <th>day_moment</th>\n",
       "      <th>...</th>\n",
       "      <th>NO3</th>\n",
       "      <th>NOX</th>\n",
       "      <th>PO4</th>\n",
       "      <th>Si</th>\n",
       "      <th>TEP</th>\n",
       "      <th>sea_surface_height_above_sea_level</th>\n",
       "      <th>surface_baroclinic_sea_water_velocity</th>\n",
       "      <th>PP</th>\n",
       "      <th>Phaeocystis_total_tpm</th>\n",
       "      <th>TPM_standardized</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>c_000002143804</td>\n",
       "      <td>adh_short</td>\n",
       "      <td>51_21</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>51</td>\n",
       "      <td>21</td>\n",
       "      <td>51.531851</td>\n",
       "      <td>3.182475</td>\n",
       "      <td>2023-04-19 09:03:00</td>\n",
       "      <td>Day</td>\n",
       "      <td>...</td>\n",
       "      <td>9.13</td>\n",
       "      <td>9.35</td>\n",
       "      <td>0.17</td>\n",
       "      <td>6.98</td>\n",
       "      <td>639.653333</td>\n",
       "      <td>-0.751302</td>\n",
       "      <td>0.899872</td>\n",
       "      <td>48.300</td>\n",
       "      <td>274.394819</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c_000002143804</td>\n",
       "      <td>adh_short</td>\n",
       "      <td>51_7</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>51</td>\n",
       "      <td>7</td>\n",
       "      <td>51.532258</td>\n",
       "      <td>3.182319</td>\n",
       "      <td>2023-04-18 19:03:00</td>\n",
       "      <td>Civil twilight</td>\n",
       "      <td>...</td>\n",
       "      <td>7.89</td>\n",
       "      <td>8.07</td>\n",
       "      <td>0.16</td>\n",
       "      <td>7.40</td>\n",
       "      <td>201.302667</td>\n",
       "      <td>-1.046195</td>\n",
       "      <td>1.005741</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1612.400269</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>c_000002143804</td>\n",
       "      <td>adh_short</td>\n",
       "      <td>51_11</td>\n",
       "      <td>3.78494</td>\n",
       "      <td>51</td>\n",
       "      <td>11</td>\n",
       "      <td>51.531646</td>\n",
       "      <td>3.182801</td>\n",
       "      <td>2023-04-18 23:07:00</td>\n",
       "      <td>Night</td>\n",
       "      <td>...</td>\n",
       "      <td>7.22</td>\n",
       "      <td>7.40</td>\n",
       "      <td>0.13</td>\n",
       "      <td>7.40</td>\n",
       "      <td>144.862667</td>\n",
       "      <td>1.264855</td>\n",
       "      <td>0.485325</td>\n",
       "      <td>34.152</td>\n",
       "      <td>1789.126177</td>\n",
       "      <td>0.002116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>c_000002143804</td>\n",
       "      <td>adh_short</td>\n",
       "      <td>51_8</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>51</td>\n",
       "      <td>8</td>\n",
       "      <td>51.531966</td>\n",
       "      <td>3.182484</td>\n",
       "      <td>2023-04-18 20:04:00</td>\n",
       "      <td>Nautical twilight</td>\n",
       "      <td>...</td>\n",
       "      <td>7.89</td>\n",
       "      <td>8.08</td>\n",
       "      <td>0.15</td>\n",
       "      <td>7.34</td>\n",
       "      <td>161.794667</td>\n",
       "      <td>-0.643491</td>\n",
       "      <td>0.946217</td>\n",
       "      <td>51.972</td>\n",
       "      <td>5576.978714</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>c_000002143804</td>\n",
       "      <td>adh_short</td>\n",
       "      <td>51_10</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>51.531665</td>\n",
       "      <td>3.182699</td>\n",
       "      <td>2023-04-18 22:05:00</td>\n",
       "      <td>Night</td>\n",
       "      <td>...</td>\n",
       "      <td>7.27</td>\n",
       "      <td>7.44</td>\n",
       "      <td>0.13</td>\n",
       "      <td>7.21</td>\n",
       "      <td>212.590667</td>\n",
       "      <td>0.313989</td>\n",
       "      <td>0.422929</td>\n",
       "      <td>31.584</td>\n",
       "      <td>2078.523171</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    transcript_id      PFAMs sample      TPM  StationPrefix  StationSuffix  \\\n",
       "0  c_000002143804  adh_short  51_21  0.00000             51             21   \n",
       "1  c_000002143804  adh_short   51_7  0.00000             51              7   \n",
       "2  c_000002143804  adh_short  51_11  3.78494             51             11   \n",
       "3  c_000002143804  adh_short   51_8  0.00000             51              8   \n",
       "4  c_000002143804  adh_short  51_10  0.00000             51             10   \n",
       "\n",
       "    Latitude  Longitude                 Date         day_moment  ...   NO3  \\\n",
       "0  51.531851   3.182475  2023-04-19 09:03:00                Day  ...  9.13   \n",
       "1  51.532258   3.182319  2023-04-18 19:03:00     Civil twilight  ...  7.89   \n",
       "2  51.531646   3.182801  2023-04-18 23:07:00              Night  ...  7.22   \n",
       "3  51.531966   3.182484  2023-04-18 20:04:00  Nautical twilight  ...  7.89   \n",
       "4  51.531665   3.182699  2023-04-18 22:05:00              Night  ...  7.27   \n",
       "\n",
       "    NOX   PO4    Si         TEP  sea_surface_height_above_sea_level  \\\n",
       "0  9.35  0.17  6.98  639.653333                           -0.751302   \n",
       "1  8.07  0.16  7.40  201.302667                           -1.046195   \n",
       "2  7.40  0.13  7.40  144.862667                            1.264855   \n",
       "3  8.08  0.15  7.34  161.794667                           -0.643491   \n",
       "4  7.44  0.13  7.21  212.590667                            0.313989   \n",
       "\n",
       "   surface_baroclinic_sea_water_velocity      PP  Phaeocystis_total_tpm  \\\n",
       "0                               0.899872  48.300             274.394819   \n",
       "1                               1.005741   0.000            1612.400269   \n",
       "2                               0.485325  34.152            1789.126177   \n",
       "3                               0.946217  51.972            5576.978714   \n",
       "4                               0.422929  31.584            2078.523171   \n",
       "\n",
       "   TPM_standardized  \n",
       "0          0.000000  \n",
       "1          0.000000  \n",
       "2          0.002116  \n",
       "3          0.000000  \n",
       "4          0.000000  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a dataframe containing the sample names, functional annotation categories of interest and TPM\n",
    "functional_columns = ['transcript_id', 'PFAMs']\n",
    "data = functional_annotation[functional_columns].merge(tpm, on='transcript_id', how='inner')\n",
    "data = data.merge(env_data, on='sample', how='inner')\n",
    "\n",
    "# Standardize the TPM values per sample by dividing by the total TPM of Phaeocystis in that sample\n",
    "data = data.merge(phaeo_tpm_sums, on='sample', how='inner')\n",
    "data['TPM_standardized'] = data['TPM'] / data['Phaeocystis_total_tpm']\n",
    "\n",
    "# Phaeocystis is absent from samples 130_15, 130_16, 130_20, 130_24, 130_25. Set these standardized TPM values to 0 otherwise they won't make sense\n",
    "data.loc[data['sample'].isin(['130_15', '130_16', '130_20', '130_24', '130_25']), 'TPM_standardized'] = 0\n",
    "\n",
    "# Inspect the data\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I've downloaded the Pfam annotation data from [interprot](https://www.ebi.ac.uk/interpro/download/Pfam/) on 30/05/2023 and stored it in `data/annotation/functional/`.\n",
    "We'll load that dataframe and parse it, to swap the PFAM short names for their descriptions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dictionary to store PFAM ID and description\n",
    "pfam_mappping = {}\n",
    "pfam_data = {}\n",
    "\n",
    "# Read the PFAM data file\n",
    "with open(\"../data/annotation/functional/Pfam-A.hmm.dat\", \"r\") as file:\n",
    "    file_content = file.read()\n",
    "\n",
    "# Split the file content into individual PFAM entries\n",
    "pfam_entries = re.split(r\"//\\n\", file_content.strip())\n",
    "\n",
    "# Process each PFAM entry\n",
    "for entry in pfam_entries:\n",
    "    lines = entry.strip().split(\"\\n\")\n",
    "    pfam_short_name_line = lines[1]\n",
    "    pfam_id_line = lines[2]\n",
    "    pfam_description_line = lines[3]\n",
    "    \n",
    "    # Extract the PFAM short name\n",
    "    pfam_short_name = pfam_short_name_line.split(\"ID\")[1].strip()\n",
    "    \n",
    "    # Extract the PFAM ID\n",
    "    pfam_id = pfam_id_line.split(\"AC\")[1].strip()\n",
    "\n",
    "    # Extract the PFAM description\n",
    "    pfam_description = pfam_description_line.split(\"DE\")[1].strip()\n",
    "\n",
    "    # Store the PFAM ID and short name in the dictionary\n",
    "    pfam_mappping[pfam_short_name] = pfam_id\n",
    "    \n",
    "    # Store the PFAM ID and description in the dictionary\n",
    "    pfam_data[pfam_id] = pfam_description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add description and ID columns to data\n",
    "data['PFAMs_ID'] = data['PFAMs'].apply(lambda x: \",\".join([pfam_mappping.get(pfam.strip(), \"-\") for pfam in x.split(\",\")]))\n",
    "data['PFAMs_description'] = data['PFAMs_ID'].apply(lambda x: \",\".join([pfam_data.get(pfam.strip(), \"-\") for pfam in x.split(\",\")]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>PFAMs</th>\n",
       "      <th>sample</th>\n",
       "      <th>TPM</th>\n",
       "      <th>StationPrefix</th>\n",
       "      <th>StationSuffix</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>Longitude</th>\n",
       "      <th>Date</th>\n",
       "      <th>day_moment</th>\n",
       "      <th>...</th>\n",
       "      <th>PO4</th>\n",
       "      <th>Si</th>\n",
       "      <th>TEP</th>\n",
       "      <th>sea_surface_height_above_sea_level</th>\n",
       "      <th>surface_baroclinic_sea_water_velocity</th>\n",
       "      <th>PP</th>\n",
       "      <th>Phaeocystis_total_tpm</th>\n",
       "      <th>TPM_standardized</th>\n",
       "      <th>PFAMs_ID</th>\n",
       "      <th>PFAMs_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>c_000002143804</td>\n",
       "      <td>adh_short</td>\n",
       "      <td>51_21</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>51</td>\n",
       "      <td>21</td>\n",
       "      <td>51.531851</td>\n",
       "      <td>3.182475</td>\n",
       "      <td>2023-04-19 09:03:00</td>\n",
       "      <td>Day</td>\n",
       "      <td>...</td>\n",
       "      <td>0.17</td>\n",
       "      <td>6.98</td>\n",
       "      <td>639.653333</td>\n",
       "      <td>-0.751302</td>\n",
       "      <td>0.899872</td>\n",
       "      <td>48.300</td>\n",
       "      <td>274.394819</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PF00106.28</td>\n",
       "      <td>short chain dehydrogenase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c_000002143804</td>\n",
       "      <td>adh_short</td>\n",
       "      <td>51_7</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>51</td>\n",
       "      <td>7</td>\n",
       "      <td>51.532258</td>\n",
       "      <td>3.182319</td>\n",
       "      <td>2023-04-18 19:03:00</td>\n",
       "      <td>Civil twilight</td>\n",
       "      <td>...</td>\n",
       "      <td>0.16</td>\n",
       "      <td>7.40</td>\n",
       "      <td>201.302667</td>\n",
       "      <td>-1.046195</td>\n",
       "      <td>1.005741</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1612.400269</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PF00106.28</td>\n",
       "      <td>short chain dehydrogenase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>c_000002143804</td>\n",
       "      <td>adh_short</td>\n",
       "      <td>51_11</td>\n",
       "      <td>3.78494</td>\n",
       "      <td>51</td>\n",
       "      <td>11</td>\n",
       "      <td>51.531646</td>\n",
       "      <td>3.182801</td>\n",
       "      <td>2023-04-18 23:07:00</td>\n",
       "      <td>Night</td>\n",
       "      <td>...</td>\n",
       "      <td>0.13</td>\n",
       "      <td>7.40</td>\n",
       "      <td>144.862667</td>\n",
       "      <td>1.264855</td>\n",
       "      <td>0.485325</td>\n",
       "      <td>34.152</td>\n",
       "      <td>1789.126177</td>\n",
       "      <td>0.002116</td>\n",
       "      <td>PF00106.28</td>\n",
       "      <td>short chain dehydrogenase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>c_000002143804</td>\n",
       "      <td>adh_short</td>\n",
       "      <td>51_8</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>51</td>\n",
       "      <td>8</td>\n",
       "      <td>51.531966</td>\n",
       "      <td>3.182484</td>\n",
       "      <td>2023-04-18 20:04:00</td>\n",
       "      <td>Nautical twilight</td>\n",
       "      <td>...</td>\n",
       "      <td>0.15</td>\n",
       "      <td>7.34</td>\n",
       "      <td>161.794667</td>\n",
       "      <td>-0.643491</td>\n",
       "      <td>0.946217</td>\n",
       "      <td>51.972</td>\n",
       "      <td>5576.978714</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PF00106.28</td>\n",
       "      <td>short chain dehydrogenase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>c_000002143804</td>\n",
       "      <td>adh_short</td>\n",
       "      <td>51_10</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "      <td>51.531665</td>\n",
       "      <td>3.182699</td>\n",
       "      <td>2023-04-18 22:05:00</td>\n",
       "      <td>Night</td>\n",
       "      <td>...</td>\n",
       "      <td>0.13</td>\n",
       "      <td>7.21</td>\n",
       "      <td>212.590667</td>\n",
       "      <td>0.313989</td>\n",
       "      <td>0.422929</td>\n",
       "      <td>31.584</td>\n",
       "      <td>2078.523171</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PF00106.28</td>\n",
       "      <td>short chain dehydrogenase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    transcript_id      PFAMs sample      TPM  StationPrefix  StationSuffix  \\\n",
       "0  c_000002143804  adh_short  51_21  0.00000             51             21   \n",
       "1  c_000002143804  adh_short   51_7  0.00000             51              7   \n",
       "2  c_000002143804  adh_short  51_11  3.78494             51             11   \n",
       "3  c_000002143804  adh_short   51_8  0.00000             51              8   \n",
       "4  c_000002143804  adh_short  51_10  0.00000             51             10   \n",
       "\n",
       "    Latitude  Longitude                 Date         day_moment  ...   PO4  \\\n",
       "0  51.531851   3.182475  2023-04-19 09:03:00                Day  ...  0.17   \n",
       "1  51.532258   3.182319  2023-04-18 19:03:00     Civil twilight  ...  0.16   \n",
       "2  51.531646   3.182801  2023-04-18 23:07:00              Night  ...  0.13   \n",
       "3  51.531966   3.182484  2023-04-18 20:04:00  Nautical twilight  ...  0.15   \n",
       "4  51.531665   3.182699  2023-04-18 22:05:00              Night  ...  0.13   \n",
       "\n",
       "     Si         TEP  sea_surface_height_above_sea_level  \\\n",
       "0  6.98  639.653333                           -0.751302   \n",
       "1  7.40  201.302667                           -1.046195   \n",
       "2  7.40  144.862667                            1.264855   \n",
       "3  7.34  161.794667                           -0.643491   \n",
       "4  7.21  212.590667                            0.313989   \n",
       "\n",
       "   surface_baroclinic_sea_water_velocity      PP  Phaeocystis_total_tpm  \\\n",
       "0                               0.899872  48.300             274.394819   \n",
       "1                               1.005741   0.000            1612.400269   \n",
       "2                               0.485325  34.152            1789.126177   \n",
       "3                               0.946217  51.972            5576.978714   \n",
       "4                               0.422929  31.584            2078.523171   \n",
       "\n",
       "   TPM_standardized    PFAMs_ID          PFAMs_description  \n",
       "0          0.000000  PF00106.28  short chain dehydrogenase  \n",
       "1          0.000000  PF00106.28  short chain dehydrogenase  \n",
       "2          0.002116  PF00106.28  short chain dehydrogenase  \n",
       "3          0.000000  PF00106.28  short chain dehydrogenase  \n",
       "4          0.000000  PF00106.28  short chain dehydrogenase  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Predict trophic mode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Indicate where to look for the machine learning modules so we can import them into the notebook\n",
    "sys.path.append('../data/trophic-mode-ml/model/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the data for the machine learning model\n",
    "df = data[['sample', 'PFAMs_ID', 'TPM_standardized']]\n",
    "df = df.explode('PFAMs_ID')\n",
    "# Group by sample and PFAM ID and sum the TPM values\n",
    "df = df.groupby(['sample', 'PFAMs_ID']).sum().reset_index()\n",
    "\n",
    "# Remove all rows with no PFAM ID\n",
    "df = df[df['PFAMs_ID'] != \"-\"]\n",
    "\n",
    "# Remove the part of the PFAM ID that indicates the domain\n",
    "df['PFAMs_ID'] = df['PFAMs_ID'].apply(lambda x: x.split(\".\")[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "from predict_tm_lib import run_prediction\n",
    "\n",
    "def run_prediction_per_sample(df, train_data_path, feature_path, labels_path):\n",
    "    \"\"\" This will run the trophic mode prediction for all species in the dataset (month, station, sample, species list).\n",
    "        The results_df used for this was generated above, and only includes species that meet the models' criteria.\n",
    "\n",
    "    Args:\n",
    "        df (pandas dataframe): Should contain a column sample and species for which the prediction should be run\n",
    "        profiles (pandas dataframe): Should contain the profiles for all species in the df. Should have columns sample, species, PFAMs_ID, and TPM (normalized for total transcriptome bin size)\n",
    "        train_data_path (csv): csv data file with the training data as used in the model by the authors\n",
    "        feature_path (csv): csv data file with the features as used in the model by the authors\n",
    "        labels_path (csv): csv data file with the labels as used in the model by the authors\n",
    "\n",
    "    Returns:\n",
    "        dataframe: dataframe with a predicted trophic mode for all species. Columns are sample, species, and predicted trophic mode.\n",
    "    \"\"\"\n",
    "    # This will store all results\n",
    "    all_results = pd.DataFrame()\n",
    "    # Get unique samples\n",
    "    samples = df['sample'].unique()\n",
    "    for sample in samples:\n",
    "        # Get profile for this specie in this sample\n",
    "        profile = df[(df['sample'] == sample)][['PFAMs_ID', 'TPM_standardized']]\n",
    "        # Reshape profile\n",
    "        profile = profile.pivot_table(columns='PFAMs_ID', values='TPM_standardized', aggfunc='sum').fillna(0)\n",
    "        # Save into a csv file\n",
    "        if not os.path.exists('../data/analysis/phaeocystis_trophic_modes/'):\n",
    "            os.makedirs('../data/analysis/phaeocystis_trophic_modes/')\n",
    "        profile_path = f'../data/analysis/phaeocystis_trophic_modes/{sample}.csv'\n",
    "        profile.to_csv(profile_path)\n",
    "        # Run the prediction model\n",
    "        predictions = run_prediction(data=profile_path, train=train_data_path, feats=feature_path, labels=labels_path, out=f'../data/analysis/phaeocystis_trophic_modes/{sample}_preds.csv', rf=False)\n",
    "        # Collect predictions into a dataframe\n",
    "        p_df = pd.DataFrame(data={'sample': sample, 'prediction': predictions})\n",
    "        # Append to all_results dataframe\n",
    "        all_results = pd.concat([all_results, p_df], ignore_index=True)\n",
    "    # Return the final results\n",
    "    return all_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n",
      "/Users/michiel/gitlab/spring_campaign_2023/scripts/../data/trophic-mode-ml/model/predict_tm_lib.py:55: DataFormatWarning: Pfam profiles have less than the suggested 800 non-zero columns.\n",
      "  warnings.warn(\"Pfam profiles have less than the suggested 800 non-zero columns.\", DataFormatWarning)\n"
     ]
    }
   ],
   "source": [
    "# Run the model for station 130, otherwise the kernel will crash\n",
    "results = run_prediction_per_sample(\n",
    "    df=df,\n",
    "    train_data_path='../data/trophic-mode-ml/data/Field_training_data.csv', \n",
    "    feature_path='../data/trophic-mode-ml/data/Extracted_Pfams.csv', \n",
    "    labels_path='../data/trophic-mode-ml/data/Field_training_labels.csv'\n",
    "    );\n",
    "results.to_csv(f'../data/analysis/phaeocystis_trophic_modes/130_preds.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All predictions indicate Phaeocystis globosa was phototrophic."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
